File: gui/menu/data/duplicate.cpp
#include "duplicate.h"


MenuDataDuplicateObj::MenuDataDuplicateObj(QWidget* parent)
    : IodeSettingsDialog(parent)
{
    setupUi(this);

    completer = new IodeCompleter(false, false, COMMENTS, this);
    lineEdit_obj_name->setCompleter(completer);
    comboBox_iode_types->setCurrentIndex(0);

    QStringList listIodeType;
    for(const std::string& iode_type : v_iode_types) listIodeType << QString::fromStdString(iode_type);

    wComboIodeType = new WrapperComboBox(label_iode_types->text(), *comboBox_iode_types, REQUIRED_FIELD, listIodeType);
    wObjName = new WrapperQLineEdit(label_obj_name->text(), *lineEdit_obj_name, REQUIRED_FIELD);
    wDupObjName = new WrapperQLineEdit(label_dup_obj_name->text(), *lineEdit_dup_obj_name, REQUIRED_FIELD);

    mapFields["IodeType"]  = wComboIodeType;
    mapFields["ObjName"]    = wObjName;
    mapFields["DupObjName"] = wDupObjName;

    connect(comboBox_iode_types, &QComboBox::currentIndexChanged, lineEdit_obj_name, &IodeAutoCompleteLineEdit::setIodeType);

    className = "MENU_DATA_DUPLICATE_OBJ";
    loadSettings();
}

MenuDataDuplicateObj::~MenuDataDuplicateObj()
{
    delete wComboIodeType;
    delete wObjName;
    delete wDupObjName;

    delete completer;
}

void MenuDataDuplicateObj::duplicate()
{
    try
    {
        int iode_type = wComboIodeType->extractAndVerify();
        std::string obj_name = wObjName->extractAndVerify().toStdString();
        std::string dup_obj_name = wDupObjName->extractAndVerify().toStdString();

        switch (iode_type)
        {
        case COMMENTS:
            Comments.add(dup_obj_name, Comments.copy(obj_name));
            break;
        case EQUATIONS:
            throw std::runtime_error("Cannot duplicate an equation");
            break;     
        case IDENTITIES:
            Identities.add(dup_obj_name, Identities.copy(obj_name));
            break; 
        case LISTS:
            Lists.add(dup_obj_name, Lists.copy(obj_name));
            break; 
        case SCALARS:
            Scalars.add(dup_obj_name, Scalars.copy(obj_name));
            break; 
        case TABLES:
            Tables.add(dup_obj_name, Tables.copy(obj_name));
            break; 
        case VARIABLES:
            Variables.add(dup_obj_name, Variables.copy(obj_name));
            break; 
        default:
            break;
        }

        this->accept();
    }
    catch(const std::exception& e)
    {
        QMessageBox::warning(nullptr, "WARNING", QString(e.what()));
    }
}

void MenuDataDuplicateObj::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/iode_gui/menu/data/duplicate.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox

from iode_gui.settings import MixinSettingsDialog
from .ui_duplicate import Ui_MenuDataDuplicateObj

from copy import copy
from iode import (IodeType, comments, equations, identities, 
                  lists, scalars, tables, variables)


class MenuDataDuplicateObj(MixinSettingsDialog):
    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuDataDuplicateObj()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.v_iode_types = list(IodeType)
        v_iode_type_names = [iode_type.name for iode_type in self.v_iode_types]
        self.ui.comboBox_iode_types.addItems(v_iode_type_names)
        self.ui.comboBox_iode_types.setCurrentIndex(0)    

        self.ui.lineEdit_obj_name.handle_iode_type(IodeType.COMMENTS)        
        self.ui.lineEdit_obj_name.include_iode_command(False)        
        self.ui.lineEdit_obj_name.include_lec_functions(False)

        self.ui.comboBox_iode_types.currentIndexChanged.connect(self.update_name_completer)
        self.load_settings()

    @Slot(int)
    def update_name_completer(self, i_iode_type):
        iode_type: IodeType = self.v_iode_types[i_iode_type]
        self.ui.lineEdit_obj_name.reset_iode_types()
        self.ui.lineEdit_obj_name.handle_iode_type(iode_type)

    @Slot()
    def duplicate(self):
        try:
            iode_type: IodeType = self.v_iode_types[self.ui.comboBox_iode_types.currentIndex()]
            obj_name: str = self.ui.lineEdit_obj_name.text().strip()
            dup_obj_name: str = self.ui.lineEdit_dup_obj_name.text().strip()

            if iode_type == IodeType.COMMENTS:
                comments[dup_obj_name] = comments[obj_name]
            elif iode_type == IodeType.EQUATIONS:
                raise RuntimeError("Cannot duplicate an equation")
            elif iode_type == IodeType.IDENTITIES:
                identities[dup_obj_name] = copy(identities[obj_name])
            elif iode_type == IodeType.LISTS:
                lists[dup_obj_name] = copy(lists[obj_name])
            elif iode_type == IodeType.SCALARS:
                scalars[dup_obj_name] = copy(scalars[obj_name])
            elif iode_type == IodeType.TABLES:
                tables[dup_obj_name] = copy(tables[obj_name])
            elif iode_type == IodeType.VARIABLES:
                variables[dup_obj_name] = copy(variables[obj_name])

            self.accept()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", f"Failed to duplicate IODE {iode_type.name} "
                                f"'{obj_name}'.\n" + str(e))
